libraries/ms-helper/src/mcp2515.c

```diff
@@ -0,0 +1,240 @@
+  spi_exchange(storage->spi_port, payload, sizeof(payload), read_data, read_len);
+}
+
+static void prv_write(Mcp2515Storage *storage, uint8_t addr, uint8_t *write_data,
+                      size_t write_len) {
+  uint8_t payload[MCP2515_MAX_WRITE_BUFFER_LEN];
+  payload[0] = MCP2515_CMD_WRITE;
+  payload[1] = addr;
+  memcpy(&payload[2], write_data, write_len);
+  spi_exchange(storage->spi_port, payload, sizeof(payload), write_data, write_len);
+}
+
+// See 12.10: *addr = (data & mask) | (*addr & ~mask)
+static void prv_bit_modify(Mcp2515Storage *storage, uint8_t addr, uint8_t mask, uint8_t data) {
+  uint8_t payload[] = { MCP2515_CMD_BIT_MODIFY, addr, mask, data };
+  spi_exchange(storage->spi_port, payload, sizeof(payload), NULL, 0);
+}
+
+static uint8_t prv_read_status(Mcp2515Storage *storage) {
+  uint8_t payload[] = { MCP2515_CMD_READ_STATUS };
+  uint8_t read_data[1] = { 0 };
+  spi_exchange(storage->spi_port, payload, sizeof(payload), read_data, sizeof(read_data));
+
+  return read_data[0];
+}
+
+static void prv_handle_rx(Mcp2515Storage *storage, uint8_t int_flags) {
+  for (size_t i = 0; i < SIZEOF_ARRAY(s_rx_buffers); i++) {
+    Mcp2515RxBuffer *rx_buf = &s_rx_buffers[i];
+    if (int_flags & rx_buf->int_flag) {
+      // message RX
+
+      // Read ID
+      uint8_t id_payload[] = { MCP2515_CMD_READ_RX | rx_buf->id };
+      Mcp2515IdRegs read_id_regs = { 0 };
+      spi_exchange(storage->spi_port, id_payload, sizeof(id_payload), (uint8_t *)&read_id_regs,
+                   sizeof(read_id_regs));
+      Mcp2515Id id = {
+        .sid_0_2 = read_id_regs.sidl.sid_0_2,
+        .sidh = read_id_regs.sidh,
+        .eid0 = read_id_regs.eid0,
+        .eid8 = read_id_regs.eid8,
+        .eid_16_17 = read_id_regs.sidl.eid_16_17,
+      };
+      bool extended = read_id_regs.sidl.ide;
+      size_t dlc = read_id_regs.dlc.dlc;
+
+      if (!extended) {
+        // Standard IDs have garbage in the extended fields
+        id.raw &= 0x7FF;
+      }
+
+      uint8_t data_payload[] = { MCP2515_CMD_READ_RX | rx_buf->data };
+      uint64_t read_data = 0;
+      spi_exchange(storage->spi_port, data_payload, sizeof(data_payload), (uint8_t *)&read_data,
+                   sizeof(read_data));
+
+      if (storage->rx_cb != NULL) {
+        storage->rx_cb(id.raw, extended, read_data, dlc, storage->context);
+      }
+    }
+  }
+}
+
+static void prv_handle_error(Mcp2515Storage *storage, uint8_t int_flags) {
+  if (int_flags & MCP2515_CANINT_EFLAG) {
+    // Don't bother reporting errors - just clear them
+    uint8_t clear = 0x00;
+    prv_write(storage, MCP2515_CTRL_REG_EFLG, &clear, 1);
+    prv_write(storage, MCP2515_CTRL_REG_TEC, &clear, 1);
+    prv_write(storage, MCP2515_CTRL_REG_REC, &clear, 1);
+  }
+}
+
+static void prv_handle_int(const GPIOAddress *address, void *context) {
+  Mcp2515Storage *storage = context;
+
+  uint8_t int_flags = 0;
+  prv_read(storage, MCP2515_CTRL_REG_CANINTF, &int_flags, 1);
+
+  // either RX or error
```

Comment:
Capitalization. I don't usually care but this is the only sentence starting in lowercase in this file...